### PR TITLE
Add test docker image

### PR DESCRIPTION
**Pull request by mandrews**
- Add a Docker image that includes Huginn test dependencies
- Ensure specs pass in Huginn test container 

If this image were to be accepted and published, a separate `docker/test/docker-compose.yml` could be added to pull in this image.

Previous PR:
https://github.com/huginn/huginn/pull/1999

See also:
https://github.com/huginn/huginn/issues/1874#issuecomment-287248255

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -27,3 +27,4 @@ deployment/cookbooks
 manifest.yml
 config/unicorn.rb
 db/schema.rb
+.local/
```
Comment (mandrews) [inline .gitignore]: an artifact generated by phantomjs

File: docker/test/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# FIXME: once image is published inherit from official image
+# FROM cantino/huginn
+FROM singleprocess_huginn_web
```
Comment (mandrews) [inline docker/test/Dockerfile]: if `scripts/setup_env` were to be included in the official image, the test image could derive from the standard image

File: config/database.yml [hunk 3]
```diff
@@ -23,6 +23,7 @@ test:
   username: <%= ENV['DATABASE_USERNAME'].presence || "root" %>
   password:  <%= ENV['DATABASE_PASSWORD'].presence || "" %>
   socket: <%= ENV['DATABASE_SOCKET'] || ["/var/run/mysqld/mysqld.sock", "/opt/local/var/run/mysql5/mysqld.sock", "/tmp/mysql.sock"].find { |path| File.exist? path } %>
+  host: <%= ENV['DATABASE_HOST'] || "" %>
```
Comment (mandrews) [inline config/database.yml]: Allows tests to be run against non local database (see also: https://github.com/huginn/huginn/issues/1979)

File: docker/test/develop.yml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      TEST_DATABASE_NAME: huginn
+      DATABASE_USERNAME: huginn
+      DATABASE_PASSWORD: myhuginnpassword
+      MYSQL_PORT_3306_TCP_ADDR: mysql
+      MYSQL_PORT_3306_TCP_PORT: 3306
+    links:
+      - mysql
+    volumes:
+      - ../../:/app
+      - /app/vendor/bundle
```
Comment (mandrews) [inline docker/test/develop.yml]: unfortunately [`.dockerignore` does not apply to when mouting volumes](https://github.com/docker/compose/issues/2098). Mounting a remote volume over the local mount point seems to be a [standard workaround](http://stackoverflow.com/a/37898591).

File: docker/test/Dockerfile (hunk 2 above)
Comment (dsander) [inline docker/test/Dockerfile]: Does it have to derive from `cantino/huginn` or could it also be the smaller `single-process` image?

File: docker/test/develop.yml (hunk 4 above)
Comment (dsander) [inline docker/test/develop.yml]: Neat trick, I missed the fact that we can not have the bundled gems in the docker image.

File: docker/test/Dockerfile (hunk 2 above)
Comment (mandrews) [inline docker/test/Dockerfile]: @dsander good point - it can derive from `cantino/huginn-single-process`

File: docker/test/scripts/test_env [hunk 5]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+set -e
+
+/scripts/setup_env
+source /app/.env
+
+export RAILS_ENV=test
+bundle install --with test development --path vendor/bundle -j 4
+bundle exec rake db:drop db:create db:migrate
```
Comment (dsander) [inline docker/test/scripts/test_env]: What do you think about removing `db:drop` here? In my tests the migrations took a lot of time, which got a bit annoying in consecutive runs to check if a changed fixed a single spec.
To start from scratch one could always remove the docker containers (or just the database data).

File: docker/test/README.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    docker-compose -f develop.yml run huginn_test_postgres
+    docker-compose -f develop.yml down
+
+Run a specific spec using `docker-compose`:
+    docker-compose -f develop.yml run huginn_test_postgres rspec ./spec/helpers/dot_helper_spec.rb:82
+    docker-compose -f develop.yml down
+
+## Usage
+
+*TODO*
```
Comment (dsander) [inline docker/test/README.md]: Do you plan do add more documentation? If not I think we could can the `usage` section.

File: docker/test/develop.yml [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+services:
+  mysqldata:
+    image: mysql:5.7
+    command: /bin/true
+
+  mysql:
+    image: mysql:5.7
+    volumes_from:
+      - mysqldata
+    ports:
```
Comment (dsander) [inline docker/test/develop.yml]: Can we remove the port mapping? It causes problems when mysql is running natively on the docker host.

File: spec/helpers/dot_helper_spec.rb [hunk 8]
```diff
@@ -78,22 +78,43 @@ def check
...
+              \k<bar3> \[label=bar3,tooltip="Dot \x20 Bar",URL="#{Regexp.quote(agent_path(@bar3))}"\];
+            \}
+            \z
+          }x)
+        end
+      else
+        it "generates a richer DOT script" do
+          expect(agents_dot(@agents, rich: true)).to match(%r{
+            \A
+            digraph \x20 "Agent \x20 Event \x20 Flow" \{
```
Comment (mandrews) [inline spec/helpers/dot_helper_spec.rb]: updated to make the `graph` element optional

File: docker/test/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+      libfreetype6 \
+      libfreetype6-dev \
+      libfontconfig1 \
+      libfontconfig1-dev curl && \
+    curl -Ls https://bitbucket.org/ariya/phantomjs/downloads/${PHANTOM_JS}.tar.bz2 \
+      | tar jxvf - --strip-components=2 -C /usr/local/bin/ ${PHANTOM_JS}/bin/phantomjs
+
+ENV RAILS_ENV test
+RUN chown -R huginn:huginn /app && \
+    sudo -u huginn -H echo "gem 'sqlite3', '~> 1.3.11'" >> /app/Gemfile && \
```
Comment (mandrews) [inline docker/test/Dockerfile]: good point - removed this step

File: docker/test/README.md (hunk 6 above)
Comment (mandrews) [inline docker/test/README.md]: I was thinking that this section could be filled in with instructions for pre-built images- removed for now

File: docker/test/develop.yml (hunk 7 above)
Comment (mandrews) [inline docker/test/develop.yml]: good catch - left that in by accident. removed

File: docker/test/scripts/test_env (hunk 5 above)
Comment (mandrews) [inline docker/test/scripts/test_env]: good point- removed the `db:drop`. multiple spec runs are a lot smoother now 😄

**Comment by mandrews** [comment]
@dsander 
> Do you think we should build the test image on Travis or let the users build it themselves locally?

I was thinking that users could build the image with the `docker-compose -f develop.yml build`, and **also** that we could publish this image as `cantino/huginn-test`. This would allow for a default `docker-compose.yml` that would be similar to `develop.yml`, **except** that the specs could be run from the published docker image. I was also hoping that [huginn_agent could optionally run external gem specs](https://github.com/huginn/huginn_agent/blob/df2a402b3dce904852ec15ceaebd5445b870d815/lib/huginn_agent/spec_runner.rb#L37) via a docker container at some point in the future. That way [*someone*](https://github.com/huginn/huginn/issues/1874#issuecomment-287248255) could setup CI for external huginn gems with only docker installed 😄

**Comment by mandrews** [comment]
@dsander updated the `Dockerfile` to derive from the existing `cantino/huginn-single-process`. This will temporary add [an extra layer](https://github.com/huginn/huginn/pull/2002/files#diff-a05f6da4bb5f2ebf27535fecb4b3ab64R23) that can be removed later.

**Comment by dsander** [comment]
Nice, I'll wait for a few days with the merge to give others a change to review/test the new image.

File: docker/test/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,29 @@
+FROM cantino/huginn-single-process
```
Comment (dsander) [inline docker/test/Dockerfile]: Sorry to bug you again, can you change this to `huginn/huginn-single-process`?

**Comment by mandrews** [comment]
@dsander I made the `huginn/huginn-single-process` change and squashed the commits

**Comment by dsander** [comment]
Thanks a lot @mandrews!

**Pull request merged**